### ISSUE TITLE
Numeric Strings of different bases converted to a numeric of a specific base Functions

### ISSUE THREAD

**Issue by ebremer** [w3c/sparql-12#113]
## Why?

XPath functions like xsd:integer("1434") assume decimal formatted numeric strings.  Converting a hexadecimal string cannot be done with Xpath nor native sparql functions.

## Proposed solution

add sparql functions like Long("numeric string", base) and Integer("numeric string", base) where base is the base of the numeric string.

For example:
Integer("DEADBEEF",16) returns 3,735,928,559.  Integer and Long would both be included for people who need more bits.

**Comment by ericprud** [comment]
ooo, `xsd:string(?i, 0x16)` could be similarly handy, though it raises the question of overloading `xsd:string` vs. creating an explicit `toString()` for numbers.

**Comment by lisp** [comment]
would 17.3.1's "No additional operator may yield a result that replaces any result other than a type error" leave latitude to change the existing operator signatures?

**Comment by ericprud** [comment]
I don't think that actually applies 'cause (being a lawyer) `xsd:integer("DEADBEEF",16)` is a syntax error in 1.0-1.1, so technically not SPARQL.

The intention was a form of monotonicity where someone writing a SPARQL 1.0 query with some expected results wouldn't be awakened from the dead to find that SPARQL 1.99 gave a different answer. Future versions are allowed to give MORE answers but not different ones.

## minus caveat
Since this may matter in other issues, it's worth considering negations (MINUS, OPTIONAL !BOUND..., (NOT EXISTS?)) which perform [negation as failure (NAF)](https://en.wikipedia.org/wiki/Negation_as_failure). These queries can lose results as SPARQL grows in expressivity. For instance, suppose we want to flag as suspect any records saying someone joined something before they were born:
**Data:**
``` turtle
PREFIX demo: <http://demographics.example/>
PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

_:alice
  demo:name "Alice" ;
  demo:born "2000-01-01T00:00:00Z"^^xsd:dateTime ;
  demo:join "2000-01-02T00:00:00Z"^^xsd:dateTime .

_:bob
  demo:name "Bob" ;
  demo:born "2010-01-02T00:00:00Z"^^xsd:dateTime ;
  demo:join "2010-01-01T00:00:00Z"^^xsd:dateTime .

_:carol
  demo:name "Carol" ;
  demo:born "2020-01-01"^^xsd:gYear ;
  demo:join "2020-01-02"^^xsd:gYear .
```
**Query:**
``` sparql
PREFIX demo: <http://demographics.example/>
PREFIX  xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?name (?joined < ?bdate AS ?lt)
WHERE {
   ?s demo:name ?name ; demo:born ?bdate ; demo:join ?joined .
   FILTER (?joined < ?bdate)
}
```
**Results:**
```
┌───────┬──────┐
│ ?name │ ?lt  │
│ "Bob" │ true │
└───────┴──────┘
```
If I ask the question via default negation and the SPARQL evaluator doesn't have a `<` operator for xsd:date:
``` sparql
WHERE {
   ?s demo:name ?name ; demo:born ?bdate ; demo:join ?joined .
   MINUS {
     ?s demo:born ?bdate ; demo:join ?joined .
     FILTER (?joined > ?bdate)
   }
}
```
, my **Results** include Carol who, as far as the evaluator knows does not have a joined date less than her birthdate:
```
┌─────────┬──────┐
│ ?name   │ ?lt  │
│   "Bob" │ true │
│ "Carol" │   -- │
└─────────┴──────┘
```
If I add an `<` operator for xsd:date, the evaluator now sees that the joined > bdate, which means she gets removed from the results. The NAF variant of this query can be read as "report anyone who is not currently known to have a valid joined date", which changes as we learn more about how to evaluate dates.

In short, without NAF, extending SPARQL giveth rows, while NAF taketh rows away. The 17.3.1 is our promise to that effect.
